### PR TITLE
Security enhancements to ChaCha20::SetKey and CSignatureCache::ComputeEntryECDSA

### PR DESCRIPTION
**Pull request by guidovranken**
- Add a check to ChaCha20::SetKey which asserts that the specified key size is valid. This is an inexpensive check that guards against (accidental) misuse, which could result in out-of-bound reads.
- In CSignatureCache::ComputeEntryECDSA, change the way a vector pointer is resolved to prevent invoking undefined behavior if the vector is empty.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: preconditions should be checked where possible

I take it you have started fuzzing Bitcoin Core again @guidovranken? That's very good news for Bitcoin :)

Don't forget to any fuzzing trophies to https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Fuzz-Trophies :)

File: src/script/sigcache.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ class CSignatureCache
     ComputeEntryECDSA(uint256& entry, const uint256 &hash, const std::vector<unsigned char>& vchSig, const CPubKey& pubkey) const
     {
         CSHA256 hasher = m_salted_hasher_ecdsa;
-        hasher.Write(hash.begin(), 32).Write(&pubkey[0], pubkey.size()).Write(&vchSig[0], vchSig.size()).Finalize(entry.begin());
+        hasher.Write(hash.begin(), 32).Write(&pubkey[0], pubkey.size()).Write(vchSig.data(), vchSig.size()).Finalize(entry.begin());
```
Comment (MarcoFalke) [inline src/script/sigcache.cpp:56]: Why is this changed here, but the other 55 instances are left as is? Including the one in this line and another one in this file?

See also commit 592404f03

And our developer notes:

```
- Watch out for out-of-bounds vector access. `&vch[vch.size()]` is illegal,
  including `&vch[0]` for an empty vector. Use `vch.data()` and `vch.data() +
  vch.size()` instead.

**Review by maflcko — COMMENTED** [review]
I wonder if this is better done with a scripted-diff

File: src/script/sigcache.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/script/sigcache.cpp:56]: Also, it could make sense to open two pull requests for the changes here, since they seem unrelated? (One of them is also failing the fuzz task, btw)

Comment (guidovranken) [reply src/script/sigcache.cpp:56]: > Why is this changed here, but the other 55 instances are left as is? Including the one in this line and another one in this file?

I didn't change the two instances of `&pubkey[0]` in this file because the [CPubkey subscript operator](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/pubkey.h#L114) returns a reference to a [regular array](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/pubkey.h#L55) instead of a vector, so no undefined behavior rules are violated.

I didn't change any other instances of `&vch[0]` because this instance specifically was the only one which came up with fuzzing (so far) with `-D_GLIBCXX_DEBUG`.

> Also, it could make sense to open two pull requests for the changes here, since they seem unrelated?

@laanwj suggested I create two commits in 1 PR.

> (One of them is also failing the fuzz task, btw)

It fails because it [instantiates ChaCha20 with a key that is ```17..31``` bytes large](https://github.com/bitcoin/bitcoin/blob/19a56d1519fb493c3e1bd5cad55360b6b80fa52b/src/test/fuzz/crypto_chacha20.cpp#L19-L26) whereas my change requires that it is either 16 or 32 bytes.

I can either relax the condition in ChaCha20 to allow keys ```16..32``` bytes large, or change the fuzzer to pick one of `(16, 32)`, let me know which you prefer.

**Comment by practicalswift** [comment]
> I wonder if this is better done with a scripted-diff

If anyone wants to tackle `&vch[0]`-to-`vch.data()` using a scripted-diff then this could be used as a starting point (untested beyond `make check`: careful manual review of the changes will be required obviously):

```
$ sed -i 's%&\([a-zA-Z0-9_]\+\)\[0\]%\1.data()%g' $(git grep -lE '&([a-zA-Z0-9_]+)\[0\]' -- build_msvc/testconsensus/testconsensus.cpp src/bench/block_assemble.cpp src/random.cpp src/script/sigcache.cpp src/test/compress_tests.cpp src/test/crypto_tests.cpp src/test/fuzz/util.cpp src/test/script_standard_tests.cpp src/test/script_tests.cpp src/wallet/crypter.cpp src/wallet/wallet.cpp)
```

File: src/script/sigcache.cpp (hunk 1 above)
Comment (practicalswift) [reply src/script/sigcache.cpp:56]: > I didn't change any other instances of `&vch[0]` because this instance specifically was the only one which came up with fuzzing (so far) with `-D_GLIBCXX_DEBUG`.

`-D_GLIBCXX_DEBUG` is great! :)

@guidovranken, you don't happen to have any notes saved on any tweaks you had to make to get a full `-D_GLIBCXX_DEBUG` compile including dependencies for Bitcoin Core? IIRC our Boost dependencies were a bit cumbersome to get compiled with `-D_GLIBCXX_DEBUG` at least historically.

If anyone wants to add a CI job which runs the tests (unit + functional + fuzz) under `-D_GLIBCXX_DEBUG` I'd be more than happy to review such a PR.

> I can either relax the condition in ChaCha20 to allow keys `16..32` bytes large, or change the fuzzer to pick one of `(16, 32)`, […]

Asserting the precondition (like you've already done) and modifying the fuzzer to fulfil said precondition is probably the best option.

Comment (fanquake) [reply src/script/sigcache.cpp:56]: > tweaks you had to make to get a full -D_GLIBCXX_DEBUG compile including dependencies for Bitcoin Core?
> IIRC our Boost dependencies were a bit cumbersome to get compiled with -D_GLIBCXX_DEBUG at least historically.

Note that depends builds with `DEBUG=1` have always used [`-D_GLIBCXX_DEBUG` && `D_GLIBCXX_DEBUG_PEDANTIC`](https://github.com/bitcoin/bitcoin/blob/master/depends/hosts/linux.mk#L10). If building depends with `DEBUG=1` don't work for some reason, please open an issue.

Comment (MarcoFalke) [reply src/script/sigcache.cpp:56]: Thanks @guidovranken and @fanquake for suggesting both `-D_GLIBCXX_DEBUG` and `-D_GLIBCXX_DEBUG_PEDANTIC`

Looks like there are more issues than this one. I've reported them here:
* https://github.com/boostorg/date_time/issues/190
* https://github.com/bitcoin/bitcoin/pull/21802

Comment (MarcoFalke) [reply src/script/sigcache.cpp:56]: Is there a downside splitting this into two pulls? I'd still prefer that to make it easier to review, because both commits can be reviewed independently. Also, it makes a cleaner history and discussion because two unrelated threads aren't intertwined.

Comment (MarcoFalke) [reply src/script/sigcache.cpp:56]: I went ahead and split the raw data pointer commit into a separate pull (#21817). I hope this is ok. The commit hash needed to be changed, but the author information is fully preserved.

File: src/crypto/chacha20.cpp [hunk 2]
```diff
@@ -23,6 +24,10 @@ static const unsigned char tau[] = "expand 16-byte k";
 
 void ChaCha20::SetKey(const unsigned char* k, size_t keylen)
 {
+    if (keylen != 16 && keylen != 32) {
+        throw std::invalid_argument("ChaCha20 key size must be 16 or 32");
```
Comment (MarcoFalke) [inline src/crypto/chacha20.cpp:28]: Wouldn't it be better to make this an assert? This would allow to compile the crypto code with `noexcept`

**Review by maflcko — APPROVED** [review]
ACK, but still needs the tests fixed

File: src/crypto/chacha20.cpp (hunk 2 above)
Comment (laanwj) [reply src/crypto/chacha20.cpp:28]: I too think an assert would be more appropriate here.  It's unclear how an exception will be handled. Also the caller is responsible that a valid key is passed. When this doesn't happen it is a coding bug, it makes more sense to immediately terminate the process.

**Comment by laanwj** [comment]
Concept ACK

File: src/crypto/chacha20.cpp [hunk 3]
```diff
@@ -23,6 +24,8 @@ static const unsigned char tau[] = "expand 16-byte k";
 
 void ChaCha20::SetKey(const unsigned char* k, size_t keylen)
```
Comment (jnewbery) [inline src/crypto/chacha20.cpp:25]: This function's declaration is commented:

```
    void SetKey(const unsigned char* key, size_t keylen); //!< set key with flexible keylength; 256bit recommended */
```

That should be updated to say that `keylen` is in bytes, and that only 16 and 32 are valid.

**Review by jnewbery — COMMENTED** [review]
PR description should be updated: Add a check to ChaCha20::SetKey which ~validates the size of the specified key, and throws an exception if it is invalid~ asserts that the specified key size is valid.

File: src/crypto/chacha20.cpp (hunk 3 above)
Comment (jnewbery) [reply src/crypto/chacha20.cpp:25]: The ctor which takes a keylen should also be commented to document assumptions.

**Comment by MarcoFalke** [comment]
Are there steps to reproduce the `ChaCha20` issue?

**Comment by MarcoFalke** [comment]
@guidovranken Are you still working on this?

File: src/crypto/chacha20.cpp (hunk 2 above)
Comment (guidovranken) [reply src/crypto/chacha20.cpp]: Replaced with assert.

File: src/crypto/chacha20.cpp (hunk 3 above)
Comment (guidovranken) [reply src/crypto/chacha20.cpp:25]: Done.

**Comment by guidovranken** [comment]
> PR description should be updated: Add a check to ChaCha20::SetKey which ~validates the size of the specified key, and throws an exception if it is invalid~ asserts that the specified key size is valid.

PR desc updated.

> Are there steps to reproduce the `ChaCha20` issue?

Should I add a test?

> @guidovranken Are you still working on this?

I've addressed the outstanding issues. Anything else?

**Comment by MarcoFalke** [comment]
See https://github.com/bitcoin/bitcoin/pull/21781#issuecomment-832857237

**Comment by MarcoFalke** [comment]
> Should I add a test?

For example. Any other way to first observe the bug and then to observe the fix is also welcome.

**Comment by guidovranken** [comment]
> > Should I add a test?
> 
> For example. Any other way to first observe the bug and then to observe the fix is also welcome.

How can I test if an assertion failure occurs in the tests?

**Comment by MarcoFalke** [comment]
I was wondering how you've found the bug on master. Code review or some other way?

**Comment by guidovranken** [comment]
> I was wondering how you've found the bug on master. Code review or some other way?

I found the bug with Cryptofuzz, which is now running on OSS-Fuzz as part of the `bitcoin-core` project.

In Cryptofuzz, I work around the bug by not passing an invalid key size: https://github.com/guidovranken/cryptofuzz/blob/61f27a0afa28942a8bdb8de49775cb734acf0121/modules/bitcoin/module.cpp#L252-L254

I can remove that check once this PR is merged, and add no unit tests to `bitcoin`. OK?
